### ISSUE TITLE
Confusion of using in a multi-locale environment

### ISSUE THREAD

**Issue by renyuneyun** [projectfluent/fluent-rs#64]
I guess the goal of fluent-rs is to provide a full functioning i18n system with fluent syntax, so the environment of multi-locale will be encountered.

The signature of `MessageContext::new()` takes a slice of strings as the parameter, which is said to 

> represents the best possible fallback chain for a given locale

However, in the next section:

> MessageContext stores messages in a single locale, but keeps a locale fallback chain for the purpose of language negotiation with i18n formatters. 

The design seems to be conflicting with itself - if `MessageContext` only stores value of one locale, how does it do locale fallback?

This leads to quite a confusion for me, and I don't really know how to use fluent-rs in a multi-locale environment where fallback is really needed.

**Comment by zbraniecki** [comment]
> This leads to quite a confusion for me, and I don't really know how to use fluent-rs in a multi-locale environment where fallback is really needed.

Great point! Sorry for the confusion.

The reason we take a list of locales is to allow for Intl formatters that are (or, in case of Rust, will be) used by the context to fallback as well as possible between locales.

But `fluent-rs` as of today does not provide a higher level API to fallback between locales yet. In JS we have `fluent-dom` and `fluent-react` packages which both implement a higher level API which uses asynchronous iterators to lazily fallback between locales.

See, for example the `Localization` class in fluent-dom: https://github.com/projectfluent/fluent.js/blob/master/fluent-dom/src/localization.js#L12

In Rust, we will want to implement a similar logic using some cached-iterator strategy. What I'm currently waiting for with this is to stabilize fluent-rs a bit (which I hope to do quite soon around Fluent 0.7), and for async/await to land in Rust.

Hope that helps! I'll keep the issue open to improve documentation around it.

**Comment by kazimuth** [comment]
See the work I've been doing in [djc/askama #211](https://github.com/djc/askama/pull/211/files) for one use case. There I'm taking the slightly-odd approach of baking localization files into the output executable, since it's a web framework designed to deploy to a static binary anyway.

Having support for per-message localization chains would be amazing. I was thinking about implementing it myself but if it's being worked on here than I can just wait for that.

**Comment by zbraniecki** [comment]
@renyuneyun, @kazimuth  - can you both evaluate current examples in `fluent-bundle`, `fluent-fallback` and `fluent-resmgr`? They all use fallback mechanisms based on `fluent-locale` and esp `fluent-fallback` should be a good example of how in Fluent world you can deploy a localization with fallback mechanism.

It's still super early (I'll release those higher level crates as 0.1), but should be good enough to start collecting feedback.

**Comment by kazimuth** [comment]
I ended up rolling my own implementation of `fluent-resmgr` + `fluent-fallback`, [here](https://github.com/kazimuth/askama/blob/fluent/askama_shared/src/i18n/macro_impl.rs). The main difference is that I build all the bundles i'm going to need once, in a `lazy_static`, and then iterate over them during lookup, instead of dynamically building them during lookup. This makes sense in a web server context -- we want to do as much work as we can at startup, to reduce latency during request processing.

**Comment by renyuneyun** [comment]
@zbraniecki Thanks a lot. It seems the crate is missing from crate.io , at least for fluent-resmgr.

It is not `Sync` for the moment, so I didn't really make it work...

For fluent-fallback, the document is not clear how to use it. Briefly looking at the code makes me more confused. Isn't it meant to be used in conjunction with fluent-resmgr?

**Comment by zbraniecki** [comment]
@renyuneyun @kazimuth - I just merged a change #114 which hopefully will make it much easier to work with FluentBundle.

You can now just push the `FluentResource` to it, or pass a reference/Rc/Arc and handle resource caching yourself.

I'll release `fluent-fallback` with 0.7 to handle language fallbacking.

**Comment by zbraniecki** [comment]
I just released 0.7. It can handle any `Borrow<FluentResource>` which allows you to simply store resources in the bundle, or references, or use Rc/Arc.

I also released `fluent-resmgr` which can facilitate storing resources for your bundles with just references (so, avoids Rc).

On top of that, I released initial version of `fluent-fallback` which facilitates language fallbacking allowing for convenient loading of fallback languages.

This should fix this bug! Reopen if needed!
